src/rgw/rgw_data_sync.cc

```diff
@@ -5223,71 +5202,74 @@ int RGWSyncBucketCR::operate(const DoutPrefixProvider *dpp)
 
       // if the state wasn't Incremental, take a bucket-wide lease to prevent
       // different shards from duplicating the init and full sync
-      if (bucket_status.state != BucketSyncState::Incremental) {
-        assert(!bucket_lease_cr);
-        bucket_lease_cr.reset(new RGWContinuousLeaseCR(env->async_rados, env->store, status_obj,
+      if (bucket_status.state != BucketSyncState::Incremental) { 
+
+		    yield call(new RGWReadRemoteBucketIndexLogInfoCR(sc, sync_pair.dest_bs.bucket, &info));
+        if (retcode < 0) {
+          return set_cr_error(retcode);
+        }
+
+        if (!bucket_lease_cr) {
+          bucket_lease_cr.reset(new RGWContinuousLeaseCR(env->async_rados, env->store, status_obj,
                                                        lock_name, lock_duration, this));
-        yield spawn(bucket_lease_cr.get(), false);
-        while (!bucket_lease_cr->is_locked()) {
-          if (bucket_lease_cr->is_done()) {
-            tn->log(5, "failed to take bucket lease");
-            set_status("lease lock failed, early abort");
-            drain_all();
-            return set_cr_error(bucket_lease_cr->get_ret_status());
+          yield spawn(bucket_lease_cr.get(), false);
+          while (!bucket_lease_cr->is_locked()) {
+            if (bucket_lease_cr->is_done()) {
+              tn->log(5, "failed to take bucket lease");
+              set_status("lease lock failed, early abort");
+              drain_all();
+              return set_cr_error(bucket_lease_cr->get_ret_status());
+            }
+            tn->log(5, "waiting on bucket lease");
+            yield set_sleeping(true);
           }
-          tn->log(5, "waiting on bucket lease");
-          yield set_sleeping(true);
-        }
-        // reread the status after acquiring the lock
-        yield call(new ReadCR(dpp, env->async_rados, env->svc->sysobj,
+          // reread the status after acquiring the lock
+          yield call(new ReadCR(dpp, env->async_rados, env->svc->sysobj,
                               status_obj, &bucket_status, false, &objv));
-        if (retcode < 0) {
-          bucket_lease_cr->go_down();
-          drain_all();
-          bucket_lease_cr.reset();
-          return set_cr_error(retcode);
+          if (retcode < 0) {
+				    RELEASE_LOCK(bucket_lease_cr);	
+            return set_cr_error(retcode);
+          }
         }
-      }
 
-      if (bucket_status.state == BucketSyncState::Init ||
-          bucket_status.state == BucketSyncState::Stopped) {
-        assert(bucket_lease_cr);
+        if (info.syncstopped) {
+          RELEASE_LOCK(bucket_lease_cr);
+          tn->log(20, SSTR("remote bilog indicates that sync was stopped. for: " << sync_pair.source_bs.get_key()));
+          return set_cr_done();
+			  }
```

Comment:
if we bail out early, the following check in the `test_bucket_sync_disable` test fails: `check_buckets_sync_status_obj_not_exist`
we need `InitBucketFullSyncStatusCR` to be invoked before we return from the CR, so that the per shard state objects are updated

Reply:
hmm ok, we would need the lock to call `InitBucketFullSyncStatusCR`. but maybe for this syncstopped case, we should call a different coroutine that both 1) writes Stopped to the bucket sync status object, and 2) deletes any of its per-shard status objects

Reply:
although that part 2) could end up racing with another Init that recreates the objects, so probably would require the lock..

Reply:
* we take the lock before `InitBucketFullSyncStatusCR` and release it before we return in the case of the syncstopped
* if (2) needed only for cleanup then I would prefer it would be done separately

Reply:
> if (2) needed only for cleanup then I would prefer it would be done separately

+1